### DISCUSSION TITLE
Support `Request` from `undici` in `startWorker().fetch()`

### DISCUSSION THREAD

**Discussion by ianduvall** [cloudflare/workers-sdk#13530] [Feature Requests]
### Which Cloudflare product(s) does this pertain to?

Wrangler core

### What version(s) of the tool(s) are you using?

3.12.0 [Wrangler]

### What version of Node are you using?

18.17.1

### What operating system are you using?

Mac

### Describe the Bug

Testing a worker using `unstable_dev` fails to parse URL when using a `Request` object.

Example:
```ts
// example.test.ts
import { type UnstableDevWorker, unstable_dev } from 'wrangler';

describe('worker', () => {
	let worker: UnstableDevWorker;
	beforeAll(async () => {
		worker = await unstable_dev('src/index.ts', {
			logLevel: 'debug',
			experimental: { disableExperimentalWarning: true },
		});
	});
	afterAll(async () => {
		await worker.stop();
	});

	test('example', async () => {
		const baseUrl = `http://${worker.address}:${worker.port}`;
		const url = new URL('/path', baseUrl);
		url.searchParams.set('hello', 'world');
		const request = new Request(url, {
			method: 'DELETE',
			headers: { something: 'goes here' },
		});
		console.log(url.toString(), request.url);

		// fails here
		const response = await worker.fetch(request);
		expect(response.status).toBe(404);
	});
});
```

### Please provide a link to a minimal reproduction

See example code above

### Please provide any relevant error logs

Jest output:
```
 FAIL  tests/failing.test.ts
  worker
    ✕ example (18 ms)

  ● worker › example

    TypeError: Failed to parse URL from [object Request]

      24 |
      25 |              // fails here
    > 26 |              const response = await worker.fetch(request);
         |                                            ^
      27 |              expect(response.status).toBe(404);
      28 |      });
      29 | });

      at new Request (../../node_modules/.pnpm/wrangler@3.12.0/node_modules/wrangler/wrangler-dist/cli.js:12652:19)
      at parseRequestInput (../../node_modules/.pnpm/wrangler@3.12.0/node_modules/wrangler/wrangler-dist/cli.js:152859:19)
      at Object.fetch (../../node_modules/.pnpm/wrangler@3.12.0/node_modules/wrangler/wrangler-dist/cli.js:152819:14)
      at Object.fetch (tests/failing.test.ts:26:33)

    Cause:
    TypeError: Invalid URL

      24 |
      25 |              // fails here
    > 26 |              const response = await worker.fetch(request);
         |                                            ^
      27 |              expect(response.status).toBe(404);
      28 |      });
      29 | });

      at new Request (../../node_modules/.pnpm/wrangler@3.12.0/node_modules/wrangler/wrangler-dist/cli.js:12650:25)
      at parseRequestInput (../../node_modules/.pnpm/wrangler@3.12.0/node_modules/wrangler/wrangler-dist/cli.js:152859:19)
      at Object.fetch (../../node_modules/.pnpm/wrangler@3.12.0/node_modules/wrangler/wrangler-dist/cli.js:152819:14)
      at Object.fetch (tests/failing.test.ts:26:33)
```

**Comment by forgeDeepakBalasubramanian** [comment]
As a workaround, I believe you should be able to pass in the URL and headers directly into the `fetch`
ie.
`
    const resp = await gatewayWorker.fetch("www.test.com", {
      headers: {
        Cookie: "country_code=us",
      },
    });
`

**Comment by mikedidomizio** [comment]
I believe the `Failed to parse URL from [object Request]` issue is related to [this one](https://github.com/cloudflare/miniflare/issues/454) and [this one](https://github.com/nodejs/undici/issues/2155).  This [comment](https://github.com/cloudflare/miniflare/issues/454#issuecomment-1485041451) explains the issue in detail.

This is the [line](https://github.com/cloudflare/workers-sdk/blob/main/packages/wrangler/src/api/dev.ts#L261) that gives the error, which is `Request` from `undici`.

I tried also using a `Request` from `undici@5.20.0` in my test, but had no luck.

I was curious and returned the `request` from the wrangler package like so:

```javascript
  // dev.ts

  // returns the Request object
  request: Request,
  fetch: async (input?: RequestInfo, init?: RequestInit) => {
...
```

I could therefore ensure both Request objects were equal, and in my test I tried:

```javascript
  worker = await unstable_dev("main.js", {
      experimental: { disableExperimentalWarning: true },
  });
  // create a new Request object from the Request from Wrangler
  const req = new worker.request(`http://example.com`);

  // pass their Request into them
  const resp = await worker.fetch(req);
```

and it works as expected ✅.  Anyways not a solution, but figured I'd share what I found.

(Tested via node@16.17.0 and node@18.18.2)

**Comment by penalosa** [comment]
Going forward, we'll be recommending the `unstable_startWorker()` API, which we'll be stabilising as `startWorker()`. It looks like the same bug exists there though, so I've marked this issue as an enhancement for `startWorker()`

**Comment by MridulTailor** [comment]
Hi, I'd like to take a shot at fixing this if it's still open for contributions.

From looking at the code, it looks like the root cause is in `ProxyStubHandler.#fetcherFetchCall()` in `miniflare/src/plugins/core/proxy/client.ts`, where a new `Request` is constructed from the raw fetch args. When a cross-realm `Request` (e.g. from `undici`) is passed in, it gets stringified to `"[object Request]"` instead of having its URL extracted (causes invalid URL error)

Proposed Plan:
- Normalize the input in `#fetcherFetchCall()` by checking for a `.url` property rather than using`instanceof`, so any Request-like object is handled correctly
- Add a regression test in `packages/miniflare/test/plugins/core/proxy/client.spec.ts`
- Check whether `unstable_dev().fetch()` in `packages/wrangler/src/api/dev.ts` has the same gap and apply the same fix there if so

A few questions before I start:
1. Is this still something you'd want to be worked on, or is there other issue that would cover it?
2. Does the approach above sound reasonable, or do you have a preferred direction?
3. Happy to be assigned if you'd like me to go ahead.

Thanks!

**Comment by petebacondarwin** [comment]
Please go ahead and have a go if you would like @MridulTailor - I am not sure it is enough to just extract the URL, since there will be other things such as headers that need to be passed, right?

**Reply by MridulTailor** [reply]
Thanks @petebacondarwin!

Yes, extracting just the URL wouldn't be enough. The plan is to read all relevant properties from incoming Request-like object (url, method, headers, body) and use them to create a proper internal Request, so nothing gets lost across the realm boundary.
